### PR TITLE
feat(rust!,python): Extend `datetime` expression function with time zone/time unit parameters

### PR DESCRIPTION
**Pull request by stinodego**
Closes [#7416](https://github.com/pola-rs/polars/issues/7416)

Changes:
* Add `time_unit`, `time_zone` and `use_earliest` arguments to the `datetime` expression function
* On the Rust side, move everything into the standard expression architecture.

Paging @MarcoGorelli for a review 😄

### CODE REVIEW

File: crates/polars-plan/src/dsl/function_expr/datetime.rs [hunk 1]
```diff
@@ -64,6 +64,11 @@ pub enum TemporalFunction {
         closed: ClosedWindow,
     },
     Combine(TimeUnit),
+    #[cfg(feature = "timezones")]
+    DatetimeFunction {
```
Comment (stinodego) [inline crates/polars-plan/src/dsl/function_expr/datetime.rs:68]: Not the ideal name here - but `Datetime` is taken by the `datetime` _method_. We should probably do a refactor to separate the methods and functions, but let's save that for another PR.

File: crates/polars-plan/src/dsl/function_expr/temporal.rs [hunk 2]
```diff
@@ -6,6 +6,100 @@ use polars_time::prelude::*;
 
 use super::*;
 
+#[cfg(feature = "timezones")]
```
Comment (stinodego) [inline crates/polars-plan/src/dsl/function_expr/temporal.rs:9]: I'm not happy with the feature gate, but it's required to use `replace_time_zone`. Is there any way around this?

**Comment by ritchie46** [comment]
Wow! TIL about `rust!` :exploding_head:

File: crates/polars-plan/src/dsl/function_expr/temporal.rs (hunk 2 above)
Comment (MarcoGorelli) [reply crates/polars-plan/src/dsl/function_expr/temporal.rs:9]: a pattern like this is used elsewhere in the code, perhaps that works here?

https://github.com/pola-rs/polars/blob/128803b237dc13d0522c22dbccae1257ae30477e/crates/polars-core/src/chunked_array/cast.rs#L45-L52

**Comment by stinodego** [comment]
> Wow! TIL about `rust!` 🤯

I added this because a lot of our changes break Rust but not Python. Recently, Rust Polars `0.31.1` was released, which should really have been `0.32.0` as it has plenty of changes that could be considered breaking.

So yes, you can add `!` to a scope now :) that should make the Rust changelogs more useful.

File: crates/polars-plan/src/dsl/function_expr/temporal.rs (hunk 2 above)
Comment (stinodego) [reply crates/polars-plan/src/dsl/function_expr/temporal.rs:9]: Thanks, I'll look into this!

**Comment by ritchie46** [comment]
> Recently, Rust Polars 0.31.1 was released, which should really have been 0.32.0 as it has plenty of changes that could be considered breaking.

No, the breaking release was already going from 0.30 to 0.31 :)  The 0.31.1 release was only an attempt to build the docs. 

But, I do agree that we must have the breaking changes in the changelog.

**Review by MarcoGorelli — APPROVED** [review]
Nice, no objections!

**Pull request merged**